Move the Food ME to the Item ME

Goldorion opened the pull request:
This PR moves food-related parameters to a new page in the item mod element.

- [x] UI changes
- [x] Generator changes
- [x] Tests
- [x] Converter for old workspace

Note: I decided to keep the food MET variable and the class used by GSON and generators because it makes the job much easier to convert old workspaces. It also reduces the chances to have a problem similar to #1572.

Note 2: I tested the converter using this test workspace created on MCreator 2021.3.45419 with Forge 1.17.1. I tried to include everything that could cause a problem and the conversion worked perfectly on my side, even when the food mod element is used in a procedure. However, please test with your workspaces as I may forget some specific cases.
[FullBackup202100345419.zip](https://github.com/MCreator/MCreator/files/7536078/FullBackup202100345419.zip)

Note 3: I didn't change the ID of label texts because it would give additional and useless work to translators.

Goldorion commented:
@Defeatomizer I changed the `use_duration` UI tip to include the meaning with foods, so can you give the new Russian translation, please?

KlemenDEV reviewed (changes requested):
You have removed fluid definition too. Please do another round of testing

Defeatomizer commented on plugins/mcreator-localization/help/default/item/use_duration.md:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+This parameter controls how long the item can be used in one take.
+
+If the item is a food, this value will define the time in ticks that the player takes to eat one item of the food.
```
Russian translation (indent as in this default tip please): Если предмет является едой, эта опция задаёт время (в тиках), необходимое игроку для употребления этой еды.

Defeatomizer commented on plugins/generator-1.17.1/forge-1.17.1/templates/item/item.java.ftl:
```diff
@@ -52,10 +52,25 @@ import net.minecraft.world.entity.ai.attributes.Attributes;
 				</#if>
 				.rarity(Rarity.${data.rarity})
+				<#if data.isFood>
+				    .food((new FoodProperties.Builder())
+				        .nutrition(${data.nutritionalValue})
+				        .saturationMod(${data.saturation}f)
+                        <#if data.isAlwaysEdible>.alwaysEat()</#if>
+                    	<#if data.forDogs>.meat()</#if>
+                    	.build())
```
Consider improving those indentations (same for [1.16.5](https://github.com/MCreator/MCreator/pull/2002/files#diff-de4152b14b40abebf0564d8e7738ae71ce523051c716456217eeca4a431785e2R114-R117)):
```suggestion
				.food((new FoodProperties.Builder())
						.nutrition(${data.nutritionalValue})
						.saturationMod(${data.saturation}f)
						<#if data.isAlwaysEdible>.alwaysEat()</#if>
						<#if data.forDogs>.meat()</#if>
						.build())
```

Defeatomizer commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1864,17 +1864,11 @@ elementgui.fluid.special_information=<html>Special information about the bucket:
 elementgui.fluid.event_flow_condition=Additional flowing condition
 elementgui.fluid.event_before_replacing_block=Before replacing a block
 elementgui.food.event_on_eaten=When food eaten
```
We don't have food mod element here anymore, rename those properties from `elementgui.food.???` to `elementgui.item.food_???`

Goldorion replied:
Note 3: I didn't change the ID of label texts because it would give additional and useless work to translators.

Defeatomizer replied:
I'd say keeping the _food_ namespace is a dis-informative thing since you removed the corresponding mod element type definition.

Goldorion replied:
@KlemenDEV should we keep `food` or change them for `item`.

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemGUI.java:
```diff
@@ -397,6 +462,26 @@ public ItemGUI(MCreator mcreator, ModElement modElement, boolean editingMode) {
 	}
 
+	private void updateFoodPanel() {
+		if(isFood.isSelected()) {
```
Very minor:
```suggestion
		if (isFood.isSelected()) {
```

Defeatomizer commented:
I'm also not sure what is gonna happen if converter fails to process food mod element. For now, it just warns that given mod element was not updated, so ISTM the workspace is gonna have unserialized ME instances...

Goldorion commented:
> 
> 
> I'm also not sure what is gonna happen if converter fails to process food mod element. For now, it just warns that given mod element was not updated, so ISTM the workspace is gonna have unserialized ME instances...

I copied the code used by the 23rd converter (fv23) and I forgot to handle this case. However, what would be the best thing to make? Should the food mod element be simply removed and potentially cause some problems? Also, it catches all exceptions, but it may only cause a real problem when MCreator deletes the food ME. The rest of the code set values using the food ME's values and it adds a new mod element. I don't see where it could have a problem.

AlsoSomeoneElse commented on src/main/java/net/mcreator/ui/modgui/ItemGUI.java:
```diff
@@ -115,6 +115,17 @@
+	private final JCheckBox forDogs = L10N.checkbox("elementgui.common.enable");
+	private final JCheckBox isAlwaysEdible = L10N.checkbox("elementgui.common.enable");
+	private ProcedureSelector onEaten;
+	private final JComboBox<String> animation = new JComboBox<>(
```
The use animation and the "On eaten" procedure could be extended to any item, not necessarily just food ("On eaten" would be "On player finishes using item")

KlemenDEV commented:
This PR has some conflicts to solve

Defeatomizer commented on src/main/java/net/mcreator/element/converter/fv26/FoodToItemConverter.java:
```diff
@@ -1,3 +1,22 @@
+
 /*
  * MCreator (https://mcreator.net/)
  * Copyright (C) 2012-2020, Pylo
```
Why 2 of those?

Goldorion replied:
Surely a bug of Intellij when I duplicated a class.

Defeatomizer commented on src/main/java/net/mcreator/ui/modgui/ItemGUI.java:
```diff
@@ -161,8 +161,8 @@ public ItemGUI(MCreator mcreator, ModElement modElement, boolean editingMode) {
-		onEaten = new ProcedureSelector(this.withEntry("item/when_eaten"), mcreator,
-				L10N.t("elementgui.food.event_on_eaten"),
+		onFinishUsingItem = new ProcedureSelector(this.withEntry("item/when_stopped_using"), mcreator,
+				L10N.t("trigger.player_useitem_finish"),
```
`trigger.*` strings are used for procedure triggers (inside Blockly), you should switch to some other key here.

Goldorion replied:
Blockly triggers use `blockly.trigger.*`. Also, even if it uses a Blockly trigger text, the line describes perfectly this trigger.

KlemenDEV replied:
It should follow elementgui.item. naming convention, though @Goldorion

KlemenDEV commented:
There is a merge conflict to be solved

Goldorion commented:
> There is a merge conflict to be solved

Done

KlemenDEV commented on src/main/java/net/mcreator/element/ModElementTypeLoader.java:
```diff
@@ -52,13 +52,11 @@ public static void loadModElements() {
 				new ModElementType<>("fuel", null, BaseType.FUEL, RecipeType.NONE, FuelGUI::new, Fuel.class));
 		ModElementType.FUNCTION = register(
-				new ModElementType<>("function", null, BaseType.DATAPACK, RecipeType.NONE, FunctionGUI::new,
+				new ModElementType<>("function", 'u', BaseType.DATAPACK, RecipeType.NONE, FunctionGUI::new,
```
u is already in use for fluid, did you mean f?

Goldorion replied:
Nope, I also changed the fluid shortcuts, so it uses `f`. I thought the fluid ME was used more often, so I changed its shortcuts for its first letter.
![image](https://user-images.githubusercontent.com/38427877/158221789-85679250-039a-4760-93c0-34b01e222c66.png)

KlemenDEV replied:
I see

KlemenDEV commented on src/main/java/net/mcreator/element/ModElementType.java:
```diff
@@ -164,4 +165,7 @@ public boolean hasProcedureTriggers() {
 	public static ModElementType<?> TOOL;
 
+	// legacy
+	public static final ModElementType<?> FOOD = new ModElementType<>("food", null, BaseType.ITEM, RecipeType.ITEM, null, Food.class); // we keep it with its values, so we can make a basic converter.
```
You should still init it as other elements, but mark it deprecated and annotated needed references with suppressed warnings

KlemenDEV commented on src/test/java/net/mcreator/integration/TestWorkspaceDataProvider.java:
```diff
@@ -1143,6 +1106,17 @@ private static GeneratableElement getExampleFor(ModElement modElement, Random ra
+			item.isFood = _true;
+			item.nutritionalValue = 5;
+			item.saturation = 0.8f;
+			item.forDogs = _true;
```
As we are moving elements, you should name those fields for mojmap mappings, so isMeat, not forDogs, it is a good time to do this now

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1967,6 +1960,8 @@ elementgui.item.dispense_properties=Custom dispense behavior
 elementgui.item.dispense_result_itemstack=On dispense attempt with result
 elementgui.item.dispense_result_itemstack.default=(shrink provided itemstack if success)
+elementgui.item.is_food=Is a food item?
+elementgui.item.food_properties= Food properties
```
Is this space after = needed?

Goldorion replied:
Typing error, oops.

KlemenDEV commented on plugins/mcreator-localization/lang/texts.properties:
```diff
@@ -1961,12 +1945,23 @@ elementgui.item.event_hand_tick=When item in hand tick
+elementgui.item.is_meat=Is this food meat?
+elementgui.item.is_edible=Is always edible?
+elementgui.item.item_animation=Item animation: 
+elementgui.item.food_3d_model=Food 3D model
```
Where is this used, please re-check translation keys?

Goldorion replied:
Ok, but what's the problem with `trigger.player_useitem_finish` in tests?

KlemenDEV replied:
It seems you reference it somewhere, maybe in tooltips?

Goldorion replied:
I guess this is due to other language files... Only those files still have this key...

Defeatomizer replied:
Try removing the `player_useitem_finish` global trigger.

Goldorion replied:
Oh yeah, they used the same key. This is also maybe why I didn't add another key.

Defeatomizer replied:
I told you about it [before](https://github.com/MCreator/MCreator/pull/2002#pullrequestreview-877913417)...

Goldorion replied:
I didn't interpret the comments. Sorry 😅So what did you mean?

KlemenDEV reviewed (commented):
Food class should be removed and mod element data parsed instead, so we keep as little legacy as needed

Goldorion commented:
Marking as a draft until we have a system handling GeneratableElemen being null (when removing a MET for example).

Goldorion commented:
The converter works, but I still have one problem. One the procedure (the one used in a trigger) is deleted...

Goldorion commented:
> The converter works, but I still have one problem. One the procedure (the one used in a trigger) is deleted (I also had this problem when testing #2497)... What is strange is the glow condition procedure works fine. I'm keeping it as a draft until this problem is fixed.

I made some tests and it seems the procedure executes the new catch block code inside the Generatable element. When printing some texts at different places, the procedure executes the code at `GeneratableElement #150`. Do you have an idea why it happens?

KlemenDEV commented:
> I made some tests and it seems the procedure executes the new catch block code inside the Generatable element. When printing some texts at different places, the procedure executes the code at GeneratableElement #150. Do you have an idea why it happens?

Not sure. Maybe try doing some prints, like stack trace of that catch block to see what call exactly causes this exception. Maybe procedure somehow loads food mod element, but to figure out why and how, a stacktrace would help

Goldorion commented:
> > I made some tests and it seems the procedure executes the new catch block code inside the Generatable element. When printing some texts at different places, the procedure executes the code at GeneratableElement #150. Do you have an idea why it happens?
> 
> Not sure. Maybe try doing some prints, like stack trace of that catch block to see what call exactly causes this exception. Maybe procedure somehow loads food mod element, but to figure out why and how, a stacktrace would help

I did. I tried to print as many things as I could, but the only IllegalArgumentException I got in the catch block is the food element, which is the only one that should execute the code. The procedure only has one block and its to create an explosion. Then, I linked it to the right click trigger of the food ME.

KlemenDEV commented:
> . I tried to print as many things as I could, but the only IllegalArgumentException I got in the catch block is the food element, which is the only one that should execute the code. 

If this is the case, than this statement is not true:

> the procedure executes the code at GeneratableElement https://github.com/MCreator/MCreator/pull/150. Do you have an idea why it happens?

So there has to be something else deleting the procedure, maybe set breakpoint on mod element delete method and see what calls it there

Goldorion commented:
> > . I tried to print as many things as I could, but the only IllegalArgumentException I got in the catch block is the food element, which is the only one that should execute the code.
> 
> If this is the case, than this statement is not true:
> 
> > the procedure executes the code at GeneratableElement #150. Do you have an idea why it happens?
> 
> So there has to be something else deleting the procedure, maybe set breakpoint on mod element delete method and see what calls it there

I'm getting the same thing as before. I don't get more info...

Goldorion commented:
> > . I tried to print as many things as I could, but the only IllegalArgumentException I got in the catch block is the food element, which is the only one that should execute the code.
> 
> If this is the case, than this statement is not true:
> 
> > the procedure executes the code at GeneratableElement #150. Do you have an idea why it happens?
> 
> So there has to be something else deleting the procedure, maybe set breakpoint on mod element delete method and see what calls it there

I found the cause and it was a pretty stupid one. Inside the converter, I deleted the food ME and I added the new Item ME, but just after the converter was executed, the catch block was doing the same thing. I never noticed it was made twice. 😅

KlemenDEV commented:
Hmm, somehow there are now some merge conflicts

KlemenDEV reviewed (commented):
Item animation should be on properties tab, not on food tab

Goldorion commented:
@KlemenDEV I just moved the food templates of the addons generator inside the item templates (because I forgot it). However, I can't test it because I don't have a bedrock edition account. Can you test it for me, please?

KlemenDEV commented on plugins/generator-1.18.2/forge-1.18.2/templates/item/item.java.ftl:
```diff
@@ -185,6 +199,37 @@ import javax.annotation.Nullable;
+        		<@procedureOBJToCode data.onFinishUsingItem/>
+        	</#if>
+
+        	<#if data.resultItem?? && !data.resultItem.isEmpty()>
```
This part should only happend when this item is food

KlemenDEV commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -329,6 +348,38 @@ import net.minecraft.entity.ai.attributes.Attributes;
         }
         </#if>
 
+        <#if data.isFood && (hasProcedure(data.onFinishUsingItem) || (data.resultItem?? && !data.resultItem.isEmpty()))>
```
This procedure should also work when it is not food, onFinishUsingItem UI element is present for all items

KlemenDEV commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -110,10 +110,29 @@ import net.minecraft.entity.ai.attributes.Attributes;
 			setRegistryName("${registryname}");
 		}
 
+        <#if data.animation != "eat">
```
This will set getUseAction for all cases except for eat. In case of default animation for food or non food item, this method should not be generated

Defeatomizer replied:
The default case is `"none"` IIRC?

Goldorion replied:
No, none will set no animation at all. If you check the current food template, it checks for the eat animation.

KlemenDEV replied:
Yes, but now this is normal item, so you need to handle cases for items too, not just food

KlemenDEV commented on plugins/generator-1.18.2/forge-1.18.2/templates/item/item.java.ftl:
```diff
@@ -54,9 +54,23 @@ import javax.annotation.Nullable;
 		);
 	}
 
+	<#if data.animation != "eat">
```
This will set getUseAction for all cases except for eat. In case of default animation for food or non food item, this method should not be generated

KlemenDEV reviewed (changes requested):
I fixed some problems with this PR, but noticed there are still quite some so handing over this back to the PR author to fix

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -121,7 +121,7 @@ import net.minecraft.entity.ai.attributes.Attributes;
 		}
 
-        <#if data.animation != "eat">
+        <#if (data.animation != "eat" && data.isFood) || (data.animation != "none" && !data.isFood)>
```
This can likely be simplified.
```suggestion
        <#if data.animation != (data.isFood?then("eat", "none"))>
```
Even better if you do these checks using `Item.hasDefaultAnimation()`.

KlemenDEV commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -329,6 +348,38 @@ import net.minecraft.entity.ai.attributes.Attributes;
         }
         </#if>
 
+        <#if (hasProcedure(data.onFinishUsingItem) || (data.eatResultItem?? && !data.eatResultItem.isEmpty()))>
```
If one chooses food, defines eatResultItem and then unselects food, this part will still generate, but should not, same for 1.18.2

Defeatomizer commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -109,28 +109,28 @@ import net.minecraft.entity.ai.attributes.Attributes;
 					.isImmuneToFire()
 					</#if>
-					.rarity(Rarity.${data.rarity})
+					.rarity(Rarity.${datay.rarity})
```
@KlemenDEV :man_facepalming:

KlemenDEV commented on plugins/generator-1.16.5/forge-1.16.5/templates/item.java.ftl:
```diff
@@ -329,6 +348,38 @@ import net.minecraft.entity.ai.attributes.Attributes;
+        <#if hasProcedure(data.onFinishUsingItem) || data.hasEatResultItem()>
+            @Override public ItemStack onItemUseFinish(ItemStack itemstack, World world, LivingEntity entity) {
+        	    ItemStack retval =
+        		    <#if data.eatResultItem?? && !data.eatResultItem.isEmpty()>
```
This should be data.hasEatResultItem() also, same in 1.18.x, otherwise if you select result item and disable food, this will still generate.

Please, test the code and check how it generates.

Goldorion replied:
> This should be data.hasEatResultItem() also, same in 1.18.x, otherwise if you select result item and disable food, this will still generate.
> 
> Please, test the code and check how it generates.

I tested, but I did not test the case a procedure could be selected and not the food checkbox.
